react/machines/shippingStateMachine.ts

```diff
@@ -0,0 +1,287 @@
+          },
+          { target: 'selectShippingOption' },
+        ],
+      },
+      selectAddress: {
+        initial: 'idle',
+        states: {
+          idle: {
+            always: [
+              {
+                target: 'createAddress',
+                cond: hasNoAvailableAddresses,
+              },
+              {
+                target: 'createAddress',
+                cond: isFirstPurchase,
+              },
+            ],
+            on: {
+              GO_TO_CREATE_ADDRESS: 'createAddress',
+              SUBMIT_SELECT_ADDRESS: {
+                target: 'submitting',
+                actions: 'updateAddressContext',
+              },
+            },
+          },
+          submitting: {
+            on: {
+              EDIT_ADDRESS: {
+                target: 'idle',
+                actions: 'clearRetryAddress',
+              },
+            },
+            invoke: {
+              src: 'tryToSelectAddress',
+              onDone: {
+                target: 'selectOption',
+                actions: ['updateSelectedAddress', 'clearRetryAddress'],
+              },
+              onError: {
+                target: 'error',
+                actions: 'updateRetryAddress',
+              },
+            },
+          },
+          error: {
+            on: {
+              RETRY_SELECT_ADDRESS: 'submitting',
+              EDIT_ADDRESS: {
+                target: 'idle',
+                actions: 'clearRetryAddress',
+              },
+            },
+          },
+          createAddress: {
+            type: 'final',
+          },
+          selectOption: {
+            type: 'final',
+          },
+        },
+
+        onDone: [
+          {
+            target: 'selectShippingOption',
+            cond: (_, __, meta) =>
+              meta.state.matches({ selectAddress: 'selectOption' }),
+          },
+          {
+            target: 'createAddress',
+          },
+        ],
+      },
+      selectShippingOption: {
+        initial: 'editing',
+        on: {
+          EDIT_ADDRESS: [
+            {
+              target: 'createAddress',
+              cond: isFirstPurchase,
+              actions: 'updateEditingAddressId',
+            },
+            { target: 'selectAddress' },
+          ],
+        },
+        states: {
+          idle: {
+            on: {
+              DESELECT_SHIPPING_OPTION: 'editing',
+              GO_TO_ADDRESS_STEP: 'done',
+            },
+          },
+          editing: {
+            on: {
+              SUBMIT_SELECT_SHIPPING_OPTION: 'submitting',
+            },
+          },
+          submitting: {
+            on: {
+              EDIT_ADDRESS: undefined,
+            },
+            invoke: {
+              src: 'tryToSelectShippingOption',
+              onDone: 'idle',
+            },
+          },
+          done: {
+            type: 'final',
+          },
+        },
+
+        onDone: 'done',
+      },
+      done: {
+        entry: 'goToNextStep',
```

Comment:
(focus only on the first and last highlighted lines)

Although the button label and the action is `go to address step`, it goes to `next step`. In a second purchase scenario, this can send the user straight to the payment instead of the login screen used to access the address step. Is that a desired behavior?

Reply:
yes